subxt/src/client/lightclient/background.rs

```diff
@@ -0,0 +1,428 @@
+                        );
+                    }
+                } else if let Some((sub_id_sender, sender)) = self.id_to_subscription.remove(&id) {
+                    let sub_id = result
+                        .get()
+                        .trim_start_matches('"')
+                        .trim_end_matches('"')
+                        .to_string();
+                    tracing::trace!(target: LOG_TARGET, "Received subscription ID: {}", sub_id);
+
+                    if sub_id_sender.send(Ok(result)).is_err() {
+                        tracing::warn!(
+                            target: LOG_TARGET,
+                            " Cannot send method response to id {:?}",
+                            id
+                        );
+                    } else {
+                        // Track this subscription ID if send is successful.
+                        self.subscriptions.insert(sub_id, sender);
+                    }
+                }
+            }
+            Ok(RpcResponse::Subscription { method, id, result }) => {
+                // Subxt calls into `author_submitAndWatchExtrinsic`, however the smoldot produces
+                // `{"event":"broadcasted","numPeers":1}` and `{"event":"validated"}` which are part
+                // of the RPC V2 API. Ignore those events.
+                if method == "transaction_unstable_watchEvent"
+                    && (result.to_string().contains("broadcasted")
+                        || result.to_string().contains("validated"))
+                {
+                    tracing::debug!(target: LOG_TARGET, "Ignoring notification {:?}", result);
+                    return;
+                }
+
+                if let Some(sender) = self.subscriptions.get_mut(&id) {
+                    // Send the current notification response.
+                    if sender.send(result).await.is_err() {
+                        tracing::warn!(
+                            target: LOG_TARGET,
+                            "Cannot send notification to subscription {:?}",
+                            id
+                        );
+
+                        // Remove the sender if the subscription dropped the receiver.
+                        self.subscriptions.remove(&id);
+                    }
+                }
+            }
+            Err(err) => {
+                tracing::warn!(target: LOG_TARGET, "cannot decode RPC response {:?}", err);
+            }
+        }
+    }
+
+    /// Perform the main background task:
+    /// - receiving requests from subxt RPC method / subscriptions
+    /// - provides the results from the light client back to users.
+    pub async fn start_task(
+        &mut self,
+        from_subxt: mpsc::Receiver<FromSubxt>,
+        from_node: smoldot_light::JsonRpcResponses,
+    ) {
+        let from_subxt_event = tokio_stream::wrappers::ReceiverStream::new(from_subxt);
+        let from_node_event = futures_util::stream::unfold(from_node, |mut from_node| async {
```

Comment:
is `stream::unfold` used here to make this Unpin? Or why not use https://docs.rs/smoldot-light/latest/smoldot_light/struct.JsonRpcResponses.html#method.next directly?

Reply:
Yep exactly, without the `unfold` stream the `full_node.next()` would not implement `Unpin` and we'll get an error similar

```
within `impl futures::Future<Output = std::option::Option<std::string::String>>`, the trait `Unpin` is not implemented for `[async fn body@JsonRpcResponses::next::{closure#0}]`
```